features/config/service/api/src/main/java/org/opennms/features/config/service/api/ConfigurationManagerService.java

```diff
@@ -0,0 +1,132 @@
+ * or (at your option) any later version.
+ *
+ * OpenNMS(R) is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with OpenNMS(R).  If not, see:
+ *      http://www.gnu.org/licenses/
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.features.config.service.api;
+
+import org.json.JSONObject;
+import org.opennms.features.config.dao.api.ConfigConverter;
+import org.opennms.features.config.dao.api.ConfigData;
+import org.opennms.features.config.dao.api.ConfigSchema;
+
+import java.io.IOException;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * Responsible for managing Schemas and Configurations.
+ * A Schema is a set of rules that constrain the data of a Configuration.
+ * A Configuration is the data that defines the runtime behaviour of a service together with the code of that service.
+ */
+public interface ConfigurationManagerService {
+
+    /**
+     * Register a service to use config manager
+     *
+     * @param configName
+     * @param majorVersion
+     * @param minorVersion
+     * @param patchVersion
+     * @param entityClass (Must have ValidateUsing annotation)
+     * @param <ENTITY>
+     * @throws IOException
+     * @throws ClassNotFoundException
+     */
+    <ENTITY> void registerSchema(final String configName, final int majorVersion, final int minorVersion,
+                                 final int patchVersion, Class<ENTITY> entityClass) throws IOException, ClassNotFoundException;
+
+    void registerSchema(final String configName, final int majorVersion, final int minorVersion,
+                        final int patchVersion, final ConfigConverter converter) throws IOException, ClassNotFoundException;
+
+    /**
+     * Get the registered Schema
+     *
+     * @param configName
+     * @return ConfigSchema
+     * @throws IOException
+     */
+    Optional<ConfigSchema<?>> getRegisteredSchema(String configName) throws IOException, ClassNotFoundException;
+
+    /**
+     * register a new configuration by xml
+     *
+     * @param configName
+     * @param configId
+     * @param xmlPath
+     * @throws IOException
+     */
+    void registerConfiguration(String configName, String configId, String xmlPath) throws IOException, ClassNotFoundException;
+
+    /**
+     * register a new configuration by JSONObject
+     *
+     * @param configName
+     * @param configId
+     * @param object
+     * @throws IOException
+     */
+    void registerConfiguration(String configName, String configId, JSONObject object)
```

Comment:
On this interface level: I wonder if we don't want to make this type save and use a Java class to transport the configuration instead of Json. But maybe Json is ok and Java would only complicate things...

Reply:
done now change to accept Object class